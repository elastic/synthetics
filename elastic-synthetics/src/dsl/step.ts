import { Page, BrowserContext, Browser } from 'playwright';
export type StepCallback = (page: Page, params: any, {context: BrowserContext, browser: Browser}) => void;

export class Step {
    name: string
    callback: StepCallback

    constructor(name: string, callback: StepCallback) {
        this.name = name;
        this.callback = callback;
    }
}