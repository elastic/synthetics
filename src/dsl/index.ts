export * from './journey';
export * from './step';
