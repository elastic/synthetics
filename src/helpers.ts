import { red, green, yellow, cyan } from 'kleur/colors';
import os from 'os';
import path from 'path';
import { performance } from 'perf_hooks';

export function noop() {}

export function indent(lines: string, tab = '   ') {
  return lines.replace(/^/gm, tab);
}

export const symbols = {
  warning: yellow('⚠'),
  skipped: cyan('-'),
  succeeded: green('✓'),
  failed: red('✖'),
};

export function formatError(error: Error) {
  if (!(error instanceof Error)) {
    return;
  }
  const { name, message, stack } = error;
  return { name, message, stack };
}

export function generateTempPath() {
  return path.join(os.tmpdir(), `synthetics-${process.hrtime().toString()}`);
}

/**
 * As per the timings used in the Network Events from
 * Chrome devtools protocol
 */
export function getMonotonicTime() {
  const hrTime = process.hrtime();
  return hrTime[0] * 1 + hrTime[1] / 1e9;
}

/**
 * Timestamp at which the current node process began.
 */
const processStart = performance.timeOrigin;
export function getTimestamp() {
  return (processStart + now()) * 1000;
}

/**
 * Relative current time from the start of the current node process
 */
export function now() {
  return performance.now();
}

/**
 * Execute all the callbacks in parallel using Promise.all
 */
export async function runParallel(callbacks) {
  const promises = callbacks.map(cb => cb());
  return await Promise.all(promises);
}
