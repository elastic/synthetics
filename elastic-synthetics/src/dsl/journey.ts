import { Step } from './step';

export type JourneyOptions = {
    name: string,
    id?: string
}

export type JourneyCallback = () => void;

export class Journey {
    options: JourneyOptions
    callback: JourneyCallback
    steps: Step[] = []

    constructor(options: JourneyOptions, callback: JourneyCallback) {
        this.options = options;
        this.callback = callback;
    }
}