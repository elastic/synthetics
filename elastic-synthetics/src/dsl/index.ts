import { Journey, JourneyCallback, JourneyOptions } from './journey'
import { Step, StepCallback } from './step'

class State {
    journeys: Journey[]
    currentJourney?: Journey

    constructor() {
        this.reset();
    }

    public reset() {
        this.journeys = [];
        this.currentJourney = undefined;
    }
}

export const state = new State();

export const journey = (options: JourneyOptions | string, callback: JourneyCallback) => {
    if (typeof options === 'string') {
        options = {name: options, id: options}
    }
    const j = new Journey(options, callback);
    state.journeys.push(j)
    state.currentJourney = j;

    j.callback(); // load steps
}

export const step = (name: string, callback: StepCallback) => {
    const step = new Step(name, callback);
    state.currentJourney.steps.push(step)
}