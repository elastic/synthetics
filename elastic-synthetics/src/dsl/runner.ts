import * as playwright from 'playwright';
import { EventEmitter } from 'events';
import { Journey } from './journey';
import { Step } from './step';
import { reporters } from '../reporters';

type RunOptions = {
    params: { [key: string]: any };
    environment: string;
    reporter?: 'default' | 'json';
    browserType?: string;
};

interface Events {
    start: {numJourneys: number}
    journeyStart: {journey: Journey, params: {[key: string]: any}}
    journeyEnd: {journey: Journey, params: {[key: string]: any}, elapsedMs: number, error: Error}
    stepStart: {journey: Journey, step: Step}
    stepEnd: {journey: Journey, step: Step, elapsedMs: number, error: Error, screenshot: string}
    end: {};
}

export default class Runner {
    eventEmitter = new EventEmitter;
    currentJourney?: Journey = null;
    journeys: Journey[] = [];

    addJourney(journey: Journey) {
        this.journeys.push(journey);
        this.currentJourney = journey;
    }

    addStep(step: Step) {
        if (this.currentJourney) {
            this.currentJourney.steps.push(step);
        }
    }

    emit<K extends keyof Events>(e: K, v: Events[K]) {
        this.eventEmitter.emit(e, v);
    }

    on<K extends keyof Events>(e: K, cb: (v: Events[K]) => void) {
        this.eventEmitter.on(e, cb);
    }

    async run(options: RunOptions) {
        const {
            browserType = 'chromium',
            params,
            reporter = 'default'
        } = options;
        /**
         * Set up the corresponding reporter
         */
        const Reporter = reporters[reporter];
        new Reporter(this);

        this.emit('start', {numJourneys: this.journeys.length});
        for (const journey of this.journeys) {
            let error: Error = undefined
            const journeyStarted = new Date().getTime();
            const browser = await playwright[browserType].launch({
                headless: false
            });
            const context = await browser.newContext();
            const page = await context.newPage();
            this.currentJourney = journey;
            this.emit('journeyStart', {journey, params});
            for (const step of journey.steps) {
                this.emit('stepStart', {journey, step});

                const started = new Date().getTime();
                let screenshot: string;
                try {
                    await step.callback(page, params, { context, browser });
                    await page.waitForLoadState('load');
                    screenshot = (await page.screenshot()).toString('base64');
                } catch(e) {
                    error = e;;
                    break; // Don't run anymore steps if we catch an error
                } finally {
                    const ended = new Date().getTime();
                    const elapsedMs = ended-started;
                    this.emit('stepEnd', {journey, step, elapsedMs, error, screenshot})
                }
            }
            await browser.close();

            const journeyEnded = new Date().getTime();
            const elapsedMs = journeyEnded-journeyStarted;
            this.emit('journeyEnd', {journey, params, elapsedMs, error})
        }
        this.emit('end', {});
        this.reset();
    }

    reset() {
        this.currentJourney = null;
        this.journeys = [];
    }
}
