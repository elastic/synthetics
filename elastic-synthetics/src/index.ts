require('source-map-support').install();

export * from './dsl';
export * from './runner';

import { run } from './runner' ;

if (require.main === module)  {
    run()
}
