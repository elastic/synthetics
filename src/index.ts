/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

import { runner } from './core';
import { RunOptions } from './core/runner';
import { setLogger } from './core/logger';
import { parseArgs } from './parse_args';

/* eslint-disable @typescript-eslint/no-var-requires */
require('source-map-support').install();

export async function run(options: RunOptions) {
  const cliArgs = parseArgs();
  /**
   * Use the NODE_ENV variable to control the environment if its not explicity
   * passed from either CLI or through the API
   */
  options.environment = options.environment || process.env['NODE_ENV'];
  /**
   * set up logger with appropriate file descriptor
   * to capture all the DEBUG logs when running from heartbeat
   */
  const outfd = options.outfd ?? cliArgs.outfd;
  setLogger(outfd);

  try {
    return await runner.run({
      ...options,
      headless: options.headless ?? cliArgs.headless,
      screenshots: options.screenshots ?? cliArgs.screenshots,
      dryRun: options.dryRun ?? cliArgs.dryRun,
      journeyName: options.journeyName ?? cliArgs.journeyName,
      network: options.network ?? cliArgs.network,
      pauseOnError: options.pauseOnError ?? cliArgs.pauseOnError,
      reporter: cliArgs.json && !options.reporter ? 'json' : options.reporter,
      outfd,
    });
  } catch (e) {
    console.error('Failed to run the test', e);
    process.exit(1);
  }
}

export { beforeAll, afterAll, journey, step, before, after } from './core';
