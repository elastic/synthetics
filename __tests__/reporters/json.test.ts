import fs from 'fs';
import { runner, step, journey } from '../../src/dsl';
import JSONReporter from '../../src/reporters/json';
import { generateTempPath } from '../../src/helpers';

describe('json reporter', () => {
  const dest = generateTempPath();
  afterAll(() => {
    fs.unlinkSync(dest);
  });

  it('writes each step as NDJSON to the FD', async () => {
    const { stream } = new JSONReporter(runner, { fd: fs.openSync(dest, 'w') });
    jest.useFakeTimers('modern').setSystemTime(new Date('2020-09-17'));
    const j1 = journey('j1', () => {});
    runner.emit('journey:start', {
      journey: j1,
      params: {},
    });
    runner.emit('step:end', {
      journey: j1,
      status: 'succeeded',
      step: step('s1', async () => {}),
      durationMs: 10,
      screenshot: 'dummy',
      url: 'dummy',
    });
    runner.emit('journey:end', {
      journey: j1,
      params: {},
      durationMs: 11,
      filmstrips: [
        {
          snapshot: 'dummy',
          name: 'screenshot',
          ts: 1,
        },
      ],
      networkinfo: [
        {
          request: {},
          response: {},
          isNavigationRequest: true,
        } as any,
      ],
    });
    runner.emit('end', 'done');

    /**
     * Close the underyling stream writing to FD to read all the contents
     */
    stream.end();
    await new Promise(resolve => stream.once('finish', resolve));
    const fd = fs.openSync(dest, 'r');
    const buffer = fs.readFileSync(fd, 'utf-8');
    expect(buffer.toString()).toMatchSnapshot();
  });
});
