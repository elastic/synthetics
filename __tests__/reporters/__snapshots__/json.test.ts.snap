// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json reporter formats network fields in ECS format 1`] = `
Object {
  "http": Object {
    "request": Object {
      "body": Object {
        "bytes": 0,
        "content": "",
      },
      "headers": Object {
        "accept_encoding": "gzip, deflate, br",
        "method": "GET",
        "path": "/",
        "upgrade_insecure_requests": "1",
        "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) HeadlessChrome/90.0.4392.0 Safari/537.36",
      },
      "initial_priority": "VeryHigh",
      "method": "GET",
      "mixed_content_type": "none",
      "referrer": undefined,
      "referrer_policy": "strict-origin-when-cross-origin",
      "url": "https://vigneshh.in/",
    },
    "response": Object {
      "body": Object {
        "bytes": 3329,
      },
      "connection_id": 12,
      "connection_reused": false,
      "encoded_data_length": 3329,
      "from_disk_cache": false,
      "from_prefetch_cache": false,
      "from_service_worker": false,
      "mime_type": "text/html",
      "protocol": "h2",
      "remote_i_p_address": "[2606:4700:3035::ac43:83e0]",
      "remote_port": 443,
      "response_time": 1612482095476.922,
      "security_details": Object {
        "certificate_id": 0,
        "certificate_transparency_compliance": "unknown",
        "cipher": "AES_128_GCM",
        "issuer": "Cloudflare Inc ECC CA-3",
        "key_exchange": "",
        "key_exchange_group": "X25519",
        "protocol": "TLS 1.3",
        "san_list": Array [
          [Function],
        ],
        "signed_certificate_timestamp_list": Array [],
        "subject_name": "sni.cloudflaressl.com",
        "valid_from": 1595980800,
        "valid_to": 1627560000,
      },
      "security_state": "secure",
      "status": 200,
      "status_code": 200,
      "status_text": "",
      "url": "https://vigneshh.in/",
    },
    "version": 2,
  },
  "tls": Object {
    "cipher": "AES_128_GCM_X25519",
    "server": Object {
      "x509": Object {
        "issuer": Object {
          "common_name": "Cloudflare Inc ECC CA-3",
        },
        "not_after": "2021-07-29T12:00:00.000Z",
        "not_before": "2020-07-29T00:00:00.000Z",
        "subject": Object {
          "common_name": "sni.cloudflaressl.com",
        },
      },
    },
    "version": "1.3",
    "version_protocol": "tls",
  },
  "url": "https://vigneshh.in/",
  "user_agent": Object {
    "name": "HeadlessChrome",
    "original": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) HeadlessChrome/90.0.4392.0 Safari/537.36",
    "version": "90.0.4392.0",
  },
}
`;

exports[`json reporter formats network fields in ECS format 2`] = `
Object {
  "http": Object {
    "request": Object {
      "body": Object {
        "bytes": 0,
        "content": "",
      },
      "headers": Object {
        "referer": "https://vigneshh.in/",
        "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) HeadlessChrome/90.0.4392.0 Safari/537.36",
      },
      "initial_priority": "Low",
      "method": "GET",
      "mixed_content_type": "none",
      "referrer": "https://vigneshh.in/",
      "referrer_policy": "strict-origin-when-cross-origin",
      "url": "https://vigneshh.in/static/main.js",
    },
    "response": Object {
      "body": Object {
        "bytes": 425,
      },
      "connection_id": 12,
      "connection_reused": true,
      "encoded_data_length": 425,
      "from_disk_cache": false,
      "from_prefetch_cache": false,
      "from_service_worker": false,
      "headers": Object {},
      "mime_type": "application/javascript",
      "protocol": "h2",
      "remote_i_p_address": "[2606:4700:3035::ac43:83e0]",
      "remote_port": 443,
      "response_time": 1612482095622.864,
      "security_details": Object {
        "certificate_id": 0,
        "certificate_transparency_compliance": "unknown",
        "cipher": "AES_128_GCM",
        "issuer": "Cloudflare Inc ECC CA-3",
        "key_exchange": "",
        "key_exchange_group": "X25519",
        "protocol": "TLS 1.3",
        "san_list": Array [
          [Function],
        ],
        "signed_certificate_timestamp_list": Array [],
        "subject_name": "sni.cloudflaressl.com",
        "valid_from": 1595980800,
        "valid_to": 1627560000,
      },
      "security_state": "secure",
      "status": 200,
      "status_code": 200,
      "status_text": "",
      "url": "https://vigneshh.in/static/main.js",
    },
    "version": 2,
  },
  "tls": Object {
    "cipher": "AES_128_GCM_X25519",
    "server": Object {
      "x509": Object {
        "issuer": Object {
          "common_name": "Cloudflare Inc ECC CA-3",
        },
        "not_after": "2021-07-29T12:00:00.000Z",
        "not_before": "2020-07-29T00:00:00.000Z",
        "subject": Object {
          "common_name": "sni.cloudflaressl.com",
        },
      },
    },
    "version": "1.3",
    "version_protocol": "tls",
  },
  "url": "https://vigneshh.in/static/main.js",
  "user_agent": Object {
    "name": "HeadlessChrome",
    "original": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) HeadlessChrome/90.0.4392.0 Safari/537.36",
    "version": "90.0.4392.0",
  },
}
`;

exports[`json reporter formats network fields in ECS format 3`] = `
Object {
  "http": Object {
    "request": Object {
      "body": Object {
        "bytes": 0,
        "content": "",
      },
      "has_post_data": true,
      "headers": Object {
        "content_type": "text/plain",
        "referer": "https://vigneshh.in/",
        "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) HeadlessChrome/90.0.4392.0 Safari/537.36",
      },
      "initial_priority": "High",
      "method": "POST",
      "mixed_content_type": "none",
      "referrer": "https://vigneshh.in/",
      "referrer_policy": "strict-origin-when-cross-origin",
      "url": "https://www.google-analytics.com/",
    },
    "response": undefined,
    "version": undefined,
  },
  "tls": undefined,
  "url": "https://www.google-analytics.com/",
  "user_agent": Object {
    "name": "HeadlessChrome",
    "original": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) HeadlessChrome/90.0.4392.0 Safari/537.36",
    "version": "90.0.4392.0",
  },
}
`;

exports[`json reporter writes each step as NDJSON to the FD 1`] = `
"{\\"type\\":\\"journey/register\\",\\"@timestamp\\":1600300800000000,\\"journey\\":{\\"name\\":\\"j1\\",\\"id\\":\\"j1\\"},\\"root_fields\\":{\\"os\\":{\\"platform\\":\\"darwin\\"},\\"package\\":{\\"name\\":\\"@elastic/synthetics\\",\\"version\\":\\"0.0.1\\"}},\\"package_version\\":\\"0.0.1\\"}
{\\"type\\":\\"journey/start\\",\\"@timestamp\\":1600300800000000,\\"journey\\":{\\"name\\":\\"j1\\",\\"id\\":\\"j1\\"},\\"root_fields\\":{\\"os\\":{\\"platform\\":\\"darwin\\"},\\"package\\":{\\"name\\":\\"@elastic/synthetics\\",\\"version\\":\\"0.0.1\\"}},\\"payload\\":{\\"params\\":{},\\"source\\":\\"async () => { }\\"},\\"package_version\\":\\"0.0.1\\"}
{\\"type\\":\\"step/screenshot\\",\\"@timestamp\\":1600300800000000,\\"journey\\":{\\"name\\":\\"j1\\",\\"id\\":\\"j1\\"},\\"step\\":{\\"name\\":\\"s1\\",\\"index\\":1},\\"root_fields\\":{\\"os\\":{\\"platform\\":\\"darwin\\"},\\"package\\":{\\"name\\":\\"@elastic/synthetics\\",\\"version\\":\\"0.0.1\\"}},\\"blob\\":\\"dummy\\",\\"package_version\\":\\"0.0.1\\"}
{\\"type\\":\\"step/end\\",\\"@timestamp\\":1600300800000000,\\"journey\\":{\\"name\\":\\"j1\\",\\"id\\":\\"j1\\"},\\"step\\":{\\"name\\":\\"s1\\",\\"index\\":1,\\"status\\":\\"succeeded\\"},\\"root_fields\\":{\\"os\\":{\\"platform\\":\\"darwin\\"},\\"package\\":{\\"name\\":\\"@elastic/synthetics\\",\\"version\\":\\"0.0.1\\"}},\\"payload\\":{\\"source\\":\\"async () => { }\\",\\"start\\":0,\\"end\\":10,\\"url\\":\\"dummy\\",\\"status\\":\\"succeeded\\"},\\"url\\":\\"dummy\\",\\"package_version\\":\\"0.0.1\\"}
{\\"type\\":\\"journey/network_info\\",\\"@timestamp\\":1600300800000000,\\"journey\\":{\\"name\\":\\"j1\\",\\"id\\":\\"j1\\"},\\"root_fields\\":{\\"user_agent\\":{},\\"http\\":{\\"request\\":{\\"body\\":{\\"bytes\\":0,\\"content\\":\\"\\"}}},\\"os\\":{\\"platform\\":\\"darwin\\"},\\"package\\":{\\"name\\":\\"@elastic/synthetics\\",\\"version\\":\\"0.0.1\\"}},\\"payload\\":{\\"request\\":{},\\"is_navigation_request\\":true,\\"browser\\":{}},\\"package_version\\":\\"0.0.1\\"}
{\\"type\\":\\"journey/filmstrips\\",\\"@timestamp\\":1600300800000000,\\"journey\\":{\\"name\\":\\"j1\\",\\"id\\":\\"j1\\"},\\"root_fields\\":{\\"os\\":{\\"platform\\":\\"darwin\\"},\\"package\\":{\\"name\\":\\"@elastic/synthetics\\",\\"version\\":\\"0.0.1\\"}},\\"payload\\":{\\"index\\":0,\\"startTime\\":392583.998697,\\"ts\\":392583998697},\\"blob\\":\\"dummy\\",\\"package_version\\":\\"0.0.1\\"}
{\\"type\\":\\"journey/end\\",\\"@timestamp\\":1600300800000000,\\"journey\\":{\\"name\\":\\"j1\\",\\"id\\":\\"j1\\",\\"status\\":\\"succeeded\\"},\\"root_fields\\":{\\"os\\":{\\"platform\\":\\"darwin\\"},\\"package\\":{\\"name\\":\\"@elastic/synthetics\\",\\"version\\":\\"0.0.1\\"}},\\"payload\\":{\\"start\\":0,\\"end\\":11,\\"status\\":\\"succeeded\\"},\\"package_version\\":\\"0.0.1\\"}
"
`;
